### DISCUSSION TITLE
Add gas limit multiplier parameter on usePrepareContractWrite

### DISCUSSION THREAD

**Discussion by brunowowk** [wevm/wagmi#2196] [Ideas] [answered]
Allow passing a gas limit multiplier on usePrepareContractWrite, to support use cases where there's considerable gas usage variability.

## Why?

One example: on auction contracts with extension for last-minute bids: if the bid is placed when the extension is triggered, the gas usage is considerably higher. If usePrepareContractWrite estimated the gas before that threshold, the gas limit might be too low and lead to a **failed tx**.

The existing alternatives lead to bad dev experience:
- using a static gas limit (not very straightforward or future-proof)
- calculating gas limit manually (there's no hook for that on wagmi)
- using a lower staleTime can increase the node usage considerably

## Suggested API:

`gasLimitMultiplier` prop on `usePrepareContractWrite` config

example:
```typescript
const { config, error } = usePrepareContractWrite({
    address: '0xecb504d39723b0be0e3a9aa33d646642d1051ee1',
    abi: wagmigotchiABI,
    functionName: 'feed',
    gasLimitMultiplier: 1.2,
  })
```

## The importance of this issue
There's a bunch of discussions suggesting the need for something like this:
- https://github.com/wagmi-dev/wagmi/discussions/839
- https://github.com/wagmi-dev/wagmi/discussions/1943
- https://github.com/wagmi-dev/wagmi/discussions/1750
- https://github.com/wagmi-dev/wagmi/discussions/1921

**Comment by brunowowk** [comment]
My current workaround is multiplying manually, btw:
```ts
const { config } = usePrepareContractWrite({
    address: '0xecb504d39723b0be0e3a9aa33d646642d1051ee1',
    abi: wagmigotchiABI,
    functionName: 'feed',
  })
  
  const { data, isLoading, isSuccess, write } = useContractWrite({
    ...config,
    request: {
      ...config.request,
      gasLimit: config.request.gasLimit.mul(115).div(100), // equivalent to a gasLimitMultiplier of 1.15
    },
  })
```

**Reply by pppest** [reply]
this works for you? have a similar case and tired recklesslySetUnpreparedOverrides.

**Reply by AndonMitev** [reply]
@volkyeth  how you get gasLimit from config returned by `usePrepareContractWrite`

```
const calldata = {
    abi: ReferralAbi,
    address: ReferralAddress,
    functionName: 'generateCodes',
  };

  const {
    config,
    isLoading: isLoadingSimulatedTx,
    error: errorSimulatedTx,
  } = usePrepareContractWrite(calldata);

  console.log(config);
```

gas && gasPrice are undefined

**Comment by 0x33dm** [comment]
@volkyeth did you find an elegant solution in the end?

@AndonMitev `usePrepareContractWrite` is not available anymore on wagmi 2.x I wonder if you found another way of doing it using wagmi?

**Reply by tmm** [reply, marked as answer]
https://wagmi.sh/react/guides/migrate-from-v1-to-v2#removed-prepare-hooks

**Reply by 0x33dm** [reply]
@tmm thanks for the quick reply.

I looked into those options and turns out [useSimulateContract](https://wagmi.sh/react/api/hooks/useSimulateContract) won't return the `gas / gasLimit` which is the value we're trying to fetch in order to add the extra "gas" buffer.

Later on I realised I had to use viem `encodeFunctionData` to encode the writeParams and then use that to call `useEstimateGas` from wagmi.

My solution in the end looks like this:

### useEstimateExtraGas
````ts
/**
 * Estimates gas for a transaction then adds extra amount, for instance
 * using BigInt(120) as percentage will add 20% extra gas to the estimated
 * amount
 */

import { encodeFunctionData, type WriteContractParameters } from 'viem'
import { useEstimateGas } from 'wagmi'

export function useEstimateExtraGas(
  params: WriteContractParameters,
  percentage: bigint
) {
  const data = params ? encodeFunctionData(params) : null

  const { data: estimatedGas } = useEstimateGas(
    data ? { data, to: params.address } : undefined
  )

  const estimatedWithExtra = estimatedGas
    ? (estimatedGas * percentage) / BigInt(100)
    : null

  return estimatedWithExtra
}
````
